File: Day 5/3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        res = []
        for i in range(n-2):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            if nums[i] > 0:
                break

            l,r = i+1, n -1

            while l <r:
                total = nums[i] + nums[l] + nums[r]
                if total < 0:
                    l += 1
                elif total > 0:
                    r -= 1
                else:
                    res.append([nums[i], nums[l], nums[r]])
                    # skip duplicates for l and r
                    left_val, right_val = nums[l], nums[r]
                    while l < r and nums[l] == left_val:
                        l += 1
                    while l < r and nums[r] == right_val:
                        r -= 1
        return res

File: Day 5/3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        sort(nums.begin(), nums.end());
        const int n = (int)nums.size();
        for (int i = 0; i < n - 2; ++i) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;   // skip duplicate i
            if (nums[i] > 0) break;                           // early stop

            int l = i + 1, r = n - 1;
            while (l < r) {
                long long total = (long long)nums[i] + nums[l] + nums[r];
                if (total < 0) {
                    ++l;
                } else if (total > 0) {
                    --r;
                } else {
                    res.push_back({nums[i], nums[l], nums[r]});
                    // skip duplicates on l and r
                    int lastL = nums[l], lastR = nums[r];
                    while (l < r && nums[l] == lastL) ++l;
                    while (l < r && nums[r] == lastR) --r;
                }
            }
        }
        return res;
    }
};
